### DISCUSSION TITLE
VRoid SDK 0.1.0以降の自動ログインについて

### DISCUSSION THREAD

**Discussion by NeoNusso** [pixiv/vroid-sdk-developers#65] [Q&A] [answered]
Oauth.Clientを使用したログインで
IsAccountFileExist() = true
IsAccessTokenExpired() = true
のときにLogin()を呼ぶとブラウザが開かれてしまいます。
サンプルコードに
// すでに認可済みだが期限切れの場合は再認可

と書いてあったので二回目以降はブラウザを開かずに再認可されるのかと思ったのですが、
想定した挙動でしょうか？

レガシーのAPIでは
AuthorizeWithExistAccountを使うことで自動ログインができていたので、同様の挙動を実現するにはどうすればよいでしょうか？
現在使用しているのはSDK 0.1.2です。

**Comment by nyamadan** [comment]
お待たせして申し訳ございません。こちら調査をしております。

> // すでに認可済みだが期限切れの場合は再認可

こちらサンプルのコメントとは異なり実装上は

```
IsAccountFileExist() = true
IsAccessTokenExpired() = false
```

の時のみブラウザを開き再認可を行う実装となっておりました。

> AuthorizeWithExistAccountを使うことで自動ログインができていたので、同様の挙動を実現するにはどうすればよいでしょうか？

現在 `Client.Login` メソッド内で再認可を行うべきかの判断を行っており、 `AuthorizeWithExistAccount` に相当するメソッドは `Legacy` のみに存在する状態です。

**Reply by NeoNusso** [reply]
> 現在 Client.Login メソッド内で再認可を行うべきかの判断を行っており

Client.Loginを呼び出した事でブラウザが開かれた事をスクリプト上で知る手段はありますか？
ゲーム側で認可コードを入力するUIを出す必要があると思いますので。

**Reply by nyamadan** [reply]
バージョン0.1.2では実装からの判断になりますが次の条件が  `true` であった場合はブラウザが開かれたと判断することができます。

```cs
!IsAccountFileExist() || IsAccessTokenExpired()
```

ただ、内部実装に依存した判定方法であるためバージョンアップなどで動作しなくなる可能性があります。
そのため、ブラウザが開かれたことを通知するコールバックの実装、あるいはブラウザを開く必要があるか判定するメソッドの実装を検討いたします。

**Reply by NeoNusso** [reply]
承知しました。
今は Legacy でログインしているため、そちらのメソッドが実装されたら切り替えようと思います。

**Reply by NeoNusso** [reply]
最初の件に戻りますが、
IsAccessTokenExpired() == true なら
!IsAccountFileExist() || IsAccessTokenExpired()　もtrueになるので、
やはりブラウザは開くという事ではないでしょうか？
定期的にExpired状態になるので現状だとかなりユーザー体験が悪いと思います。

**Reply by nyamadan** [reply]
> やはりブラウザは開くという事ではないでしょうか？

こちらはご指摘の通りです。現状実装ですと `IsAccessTokenExpired() == true` の際にブラウザが開かれることを止める方法はございません。

> 定期的にExpired状態になるので現状だとかなりユーザー体験が悪いと思います。

こちらもご指摘の通りだと思います。調査の結果改善の余地があることがわかりました。
改善に向けて引き続き調査を続けてまいります。改善の目途が立ちましたらお知らせいたします。

**Comment by nyamadan** [comment]
> 定期的にExpired状態になるので現状だとかなりユーザー体験が悪いと思います。

こちらについてですが、こちらが想定しているよりはるかに早くExpiredになることが判明しました。
そのため、まずはExpiredの期限を延ばす対応をを行います。次回リリースのバージョンでこちらは対応予定です。

**Reply by nyamadan** [reply, marked as answer]
> そのため、まずはExpiredの期限を延ばす対応をを行います。次回リリースのバージョンでこちらは対応予定です。

バージョン0.1.4にてExpiredの期限を延ばす対応をいたしました。 https://github.com/pixiv/vroid-sdk-developers/discussions/69

`IsAccountFileExist()` が `true` であれば ユーザーが明示的に連携を取り消すなどしない限りは当面ブラウザが開くことはない想定です。

> そのため、ブラウザが開かれたことを通知するコールバックの実装、あるいはブラウザを開く必要があるか判定するメソッドの実装を検討いたします。

上記については引き続き検討してまいります。
